solutions/devsprint-riccardo-1/OnboardingChallengeKit/Sources/OnboardingChallengeKit/Screens/List/ListViewController.swift

```diff
@@ -3,74 +3,92 @@
-        title = "Repositories"
-        setupNavigationBar()
-    }
-    
-    private func setupNavigationBar() {
-        navigationController?.navigationBar.backgroundColor = .secondarySystemBackground
-        navigationController?.navigationBar.prefersLargeTitles = true
-        navigationItem.searchController = searchController
-        navigationItem.hidesSearchBarWhenScrolling = false
-    }
-        
-    private func bindEvents() {
-        //insert the callback for the tapped cell in ListView
-    }
-    
-    private func fetchList() {
-        self.service.fetchList { items in
-            let configuration = ListViewConfiguration(listItems: items)
-            self.listView.updateView(with: configuration)
-        }
+  // MARK: Outlets
+
+  private lazy var listView: ListView = {
+    return ListView()
+  }()
+
+  private lazy var settingsButton: UIBarButtonItem = {
+    let button = UIBarButtonItem(title: "Settings", style: .plain, target: self, action: #selector(self.showSettings))
+    return button
+  }()
+
+  private lazy var searchController: UISearchController = {
+    let searchController = UISearchController(searchResultsController: nil)
+    searchController.delegate = self
+    searchController.searchBar.delegate = self
+    searchController.searchResultsUpdater = self
+    searchController.searchBar.placeholder = "Type a GitHub user name"
+    searchController.hidesNavigationBarDuringPresentation = false
+    return searchController
+  }()
+
+  private let service = Service()
+
+  // MARK: Initialization
+
+  public init() {
+    super.init(nibName: nil, bundle: nil)
+  }
+
+  required init?(coder _: NSCoder) {
+    fatalError("init(coder:) has not been implemented")
+  }
+
+  override public func loadView() {
+    self.view = self.listView
+  }
+
+  override public func viewDidLoad() {
+    super.viewDidLoad()
+    self.setupUI()
+    self.bindEvents()
+    self.fetchList()
+  }
+
+  // MARK: Actions
+
+  // insert action to settingsViewController here
+
+  // MARK: Methods
+
+  private func setupUI() {
+    title = "Repositories"
+    self.setupNavigationBar()
+  }
+
+  private func setupNavigationBar() {
+    navigationController?.navigationBar.backgroundColor = .secondarySystemBackground
+    navigationController?.navigationBar.prefersLargeTitles = true
+    navigationItem.searchController = self.searchController
+    navigationItem.hidesSearchBarWhenScrolling = false
+    navigationItem.rightBarButtonItem = self.settingsButton
+  }
+
+  private func bindEvents() {
+    // insert the callback for the tapped cell in ListView
+  }
+
+  private func fetchList() {
+    self.service.fetchList { items in
+      let configuration = ListViewConfiguration(listItems: items)
+      self.listView.updateView(with: configuration)
     }
+  }
+
+  @objc private func showSettings() {
+    navigationController?.present(
+      SettingsViewController(interactor: SettingsInteractor(bundleService: Bundle.main)),
```

Comment:
here we are coupling together the `ListViewController` with the navigation to the `SettingsViewController`. In a modularized architecture, it means that the `ListViewController` module must know about the `SettingsViewController`. That can create issues because if the `SettingsViewController` changes its interface, for example, the `ListViewController` must change as well.

This is a violation of the [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single-responsibility_principle): we don't want the ListViewController to change due to a change in the Settings.

Can you think about a way to decouple the two view controller? 



_hint_: it is similar to what we did with the `BundleService`, asking it from the outside...

Reply:
Would a router that handles routing from ListViewController to SettingsViewController be solving this violation? 

Or would you recommend a router that starts the transfer and then another router that sort of accepts it? So like two routers, one belongs to List and the other one to Settings?

Reply:
A router could work. However, we can use something more lightweight: a `Router` requires us to create an abstraction and a source view controller, on which the `Router` can invoke the `present` method. Something like:
```swift
protocol Router {
  func navigate(from viewController: UIViewcontroller)
}

struct ModalSettingsRouter: Router {
  func navigate(from viewController: UIViewController) {
    viewController.present( /* settings view controller code */)
  }
}
```

We can avoid completely to create these additional abstractions and objects if we use a simple closure.

The list view controller can take a closure called `onSettingsTapped` in the init and the `SceneDelegate` can implement the closure to present a proper `ViewController`.

The final solution will be something like this:

```swift 
class ListViewController: UIViewController {
  let onSettingsTapped: (UIViewController) -> Void

  init(onSettingsTapped: @escaping (UIViewController) -> Void) {
    self.onSettingsTapped = onSettingsTapped
  }

  @objc private func showSettings() {
    self.onSettingsTapped(self) // or we can unwrap the navigation controller
  }
}

class SceneDelegate {

  func scene(_ scene: UIScene, willConnectTo _: UISceneSession, options _: UIScene.ConnectionOptions) {
    // other part of the code
    let vc = NavigationController(rootViewController: ListViewController(onSettingsTapped: { presentingVC in
      presentingVC.present(SettingsViewController()) // and all its configurations
    }))
  }
}
```

By doing this, not only the `ListViewController` does not have to know anything about the `SettingsViewController`, but it does not have to know about all the interactions and other pieces of the pattern. Plus, we can decide how to present the SettingsViewController (push or present) without changing the implementation of the ListViewController.

What do you think? Does it make sense? Do you like it? 😊